### PR TITLE
Migrate Juxtapose to Bakery

### PR DESCRIPTION
**Pull request by davisshaver**
See https://github.com/fusioneng/shortcake-bakery/issues/44

### CODE REVIEW

File: inc/shortcodes/class-image-comparision.php [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+		$right_image = wp_get_attachment_image_src( $attrs['right'], 'large', false );
+
+		$left_caption = get_post_field( 'post_excerpt', $attrs['left'] );
+		$right_caption = get_post_field( 'post_excerpt', $attrs['right'] );
+
+		if ( ! $left_image || ! $right_image ) {
+			return;
+		}
+
+		wp_enqueue_script( 'juxtapose-js' );
```
Comment (davisshaver) [inline inc/shortcodes/class-image-comparision.php]: @danielbachhuber Any clue why this enqueuing wouldn't work? I know the init functions are run, but the enqueuing isn't working.

Comment (danielbachhuber) [inline inc/shortcodes/class-image-comparision.php]: Scripts should be registered on `wp_enqueue_scripts` action. `jquery` isn't registered until then, which means your dependency is failing.

Comment (davisshaver) [inline inc/shortcodes/class-image-comparision.php]: Danke

**Comment by davisshaver** [comment]
One implementation detail:
- In the Fusion version, we use our custom credit field, which can't be assumed to exist on Shortcake Bakery sites. What's the best way of handling this difference?

**Comment by danielbachhuber** [comment]
> What's the best way of handling this difference?

Hm. Template actions would be an easy hack, but not super glam IMHO. Maybe we should have a way of filtering the template to override?

@goldenapples opinions?

**Comment by davisshaver** [comment]
@danielbachhuber On closer review, I think the method of access is standard enough (image meta credit) that we can implement.

File: inc/shortcodes/class-image-comparision.php (hunk 1 above)
Comment (davisshaver) [inline inc/shortcodes/class-image-comparision.php]: @danielbachhuber `wp_enqueue_scripts` doesn't seem to run within the admin view.

Comment (danielbachhuber) [inline inc/shortcodes/class-image-comparision.php]: ... which is why we have https://github.com/fusioneng/fusion-theme/blob/master/inc/class-admin.php#L55

**Comment by danielbachhuber** [comment]
@davisshaver Can you make sure pull requests will merge cleanly before you assign for review?

**Comment by davisshaver** [comment]
@danielbachhuber My bad.

File: .travis.yml [hunk 2]
```diff
@@ -1,5 +1,3 @@
-sudo: false
-
```
Comment (danielbachhuber) [inline .travis.yml]: Merge conflict? These shouldn't be removed

File: inc/shortcodes/class-image-comparision.php [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+					),
+				array(
+					'label'  => 'Slider Start Position',
+					'attr'   => 'position',
+					'type'   => 'select',
+					'options' => array(
+						'center' => 'Center',
+						'mostlyleft' => 'Mostly Left',
+						'mostlyright' => 'Mostly Right',
+					),
```
Comment (danielbachhuber) [inline inc/shortcodes/class-image-comparision.php]: All of these text strings should be translatable.

File: inc/shortcodes/class-image-comparision.php [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+				),
+			),
+		);
+	}
+
+	public static function setup_actions() {
+		add_action( 'wp_enqueue_scripts', 'Shortcake_Bakery\Shortcodes\Image_Comparision::action_init_register_scripts' );
+		add_action( 'shortcode_ui_after_do_shortcode', function( $shortcode ) {
+			if ( false !== stripos( $shortcode, 'image-comparision' ) ) {
+				echo '<link rel="stylesheet" href="' . esc_url( SHORTCAKE_BAKERY_URL_ROOT . 'assets/css/juxtapose.css' ) . '"><script type="text/javascript" src="' . esc_url( get_template_directory_uri() . '/assets/vendor/juxtaposejs/juxtapose.js' ) . '"></script>';
```
Comment (danielbachhuber) [inline inc/shortcodes/class-image-comparision.php]: Do we need to load the JavaScript here too?

File: inc/shortcodes/class-image-comparision.php [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+					),
+				),
+			),
+		);
+	}
+
+	public static function setup_actions() {
+		add_action( 'wp_enqueue_scripts', 'Shortcake_Bakery\Shortcodes\Image_Comparision::action_init_register_scripts' );
+		add_action( 'shortcode_ui_after_do_shortcode', function( $shortcode ) {
+			if ( false !== stripos( $shortcode, 'image-comparision' ) ) {
```
Comment (danielbachhuber) [inline inc/shortcodes/class-image-comparision.php]: Should this check be `if ( 0 === stripos( $shortcode, '[image-comparision' ) ) {` to be more precise?

**Comment by danielbachhuber** [comment]
Few minor things, then this can go out first thing next week.

File: inc/shortcodes/class-image-comparision.php (hunk 4 above)
Comment (davisshaver) [inline inc/shortcodes/class-image-comparision.php]: @danielbachhuber Loading the javascript allows interactive preview in admin

**Comment by davisshaver** [comment]
@danielbachhuber should be all set now

File: .travis.yml [hunk 6]
```diff
@@ -18,10 +18,6 @@ matrix:
     - php: 5.3
       env: WP_VERSION=latest WP_MULTISITE=1
 
-cache:
-  - composer
-  - npm
-
```
Comment (danielbachhuber) [inline .travis.yml]: Why are you removing this?

File: inc/shortcodes/class-image-comparision.php [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+				),
+			),
+		);
+	}
+
+	public static function setup_actions() {
+		add_action( 'wp_enqueue_scripts', 'Shortcake_Bakery\Shortcodes\Image_Comparision::action_init_register_scripts' );
+		add_action( 'shortcode_ui_after_do_shortcode', function( $shortcode ) {
+			if ( false !== stripos( $shortcode, '[' . self::get_shortcode_tag() ) ) {
+				echo '<link rel="stylesheet" href="' . esc_url( SHORTCAKE_BAKERY_URL_ROOT . 'assets/css/juxtapose.css' ) . '"><script type="text/javascript" src="' . esc_url( get_template_directory_uri() . '/assets/vendor/juxtaposejs/juxtapose.js' ) . '"></script>';
```
Comment (danielbachhuber) [inline inc/shortcodes/class-image-comparision.php]: Why no script here?

Comment (davisshaver) [inline inc/shortcodes/class-image-comparision.php]: @danielbachhuber There is a script there...

File: .travis.yml (hunk 6 above)
Comment (davisshaver) [inline .travis.yml]: merge conflict, fixing

**Comment by davisshaver** [comment]
@danielbachhuber Back to you

**Pull request merged**